### PR TITLE
Issue 502 - unify wasm compile output

### PR DESCRIPTION
**Pull request by ash-tamraz**
This is in reference to [issue 502](https://github.com/CraneStation/cranelift/issues/502)

I'm just for feedback on what I have at this point. It doesn't feel like I've done much of anything, but I'm not really sure what else to do to unify these further. I'm wondering if maybe I've missed the point of the issue? Any feedback is appreciated!

### CODE REVIEW

File: src/compile.rs [hunk 1]
```diff
@@ -36,15 +55,26 @@ pub fn run(
 }
 
 fn handle_module(
+    _flag_verbose: bool,
```
Comment (bjorn3) [inline src/compile.rs]: Please remove leading `_`.

File: src/compile.rs [hunk 2]
```diff
@@ -10,8 +10,26 @@ use cranelift_reader::parse_test;
...
+    ($x: expr, $($tts:tt)*) => {
+        if $x {
+            print!($($tts)*);
+        }
+    }
+}
+
 pub fn run(
     files: Vec<String>,
+    _flag_verbose: bool,
```
Comment (bjorn3) [inline src/compile.rs]: Please remove leading `_`.

File: src/compile.rs (hunk 1 above)
Comment (ash-tamraz) [reply src/compile.rs]: When I initially added this code, I got a warning from cargo suggesting that I add the leading `_`. I've removed it and it looks like it's working fine now, though.

File: src/compile.rs (hunk 2 above)
Comment (ash-tamraz) [reply src/compile.rs]: I've removed them.

**Comment by sunfishcode** [comment]
This patch makes "compile" more consistent with "wasm", which is a good start. The other thing we'd like to do here is actually factor out parts of the code which are the same between the two, so that they can stay consistent as things evolve. Does that make sense?

**Comment by ash-tamraz** [comment]
>This patch makes "compile" more consistent with "wasm", which is a good start. The other thing we'd like to do here is actually factor out parts of the code which are the same between the two, so that they can stay consistent as things evolve. Does that make sense?

Definitely! I have two methods for unifying shared functionality between `src/compile.rs` and `src/wasm.rs` in mind: 
- Have `src/clif-util.rs` call a unified module which performs wasm or compile specific operations depending on a parameter indicating the calling subcommand (probably a boolean).
- Leave `src/clif-util.rs` as it is and have `src/compile.rs` and `src/wasm.rs` individually call functions in a module that unifies shared functionality as needed.

Which would you say is the best design for factoring out shared code? Or maybe you have suggestions for better designs for this application?

**Comment by ash-tamraz** [comment]
@sunfishcode I did some quick unification of code shared between `wasm` and `compile`. Let me know if the changes and design match what you had in mind, or if I need to make some changes.

**Comment by sunfishcode** [comment]
That's a good start! One of the next key pieces will be the code around the `compile_and_emit` calls. This is a little tricker, because the code paths aren't exactly the same right now, but wasm's `flag_check_translation` and `flag_print_size` are things that'd be nice to have in the execute path as well, which should make more things in common between the two.

**Comment by ash-tamraz** [comment]
> That's a good start! One of the next key pieces will be the code around the `compile_and_emit` calls. This is a little tricker, because the code paths aren't exactly the same right now, but wasm's `flag_check_translation` and `flag_print_size` are things that'd be nice to have in the execute path as well, which should make more things in common between the two.

Apologies for the delay! Here are some changes reflecting your feedback, finally. (:-P)

I renamed the shared file `clif_shared.rs`, as it was a little less wordy but still communicates the goal (I know the convention is to name files with a `-` as opposed to a `_`, but I couldn't figure out how to indicate to cargo that the `-` should be replaced with a `_` in in `use` statements).

`print_size` and `module_code_size` currently overlap, e.g. `cargo run wasm -v -X --target=x86_64 wasmtests/arith.wat ` triggers both. It shouldn't be difficult to resolve that, though.

**Comment by alexcrichton** [comment]
Thanks for the PR again, and as a procedural note the Cranelift repository [has now merged](https://github.com/bytecodealliance/cranelift/issues/1408) into the [wasmtime repository](https://github.com/bytecodealliance/wasmtime/tree/master/cranelift).

PRs are no longer landing in this repository, and unfortunately there's no "one button" solution to move a PR to the wasmtime repository. A [script has been prepared](https://gist.github.com/alexcrichton/8cb3f4ef7c25317ba6824ee31e3e53aa), however, to assist you in transferring this PR to the wasmtime repo. Feel free to reach out on Zulip with any questions!

**Comment by tschneidereit** [comment]
We'll archive this repository in favor of the Wasmtime repository — see Alex's comment about the merge above. I'm closing this PR now in order to have an explanation in a comment before archiving the repo auto-closes all PRs.

**Pull request closed without merging**